modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+                    e
+                );
+            }
+        }
+    }
+
+    Set<Index> maybeExecuteDownsampling(ClusterState state, DataStream dataStream, List<Index> targetIndices) {
+        Set<Index> affectedIndices = new HashSet<>();
+        Metadata metadata = state.metadata();
+        for (Index index : targetIndices) {
+            if (dataStream.isIndexManagedByDataStreamLifecycle(index, metadata::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                    index,
+                    metadata::index,
+                    nowSupplier
+                );
+                if (downsamplingRounds.isEmpty()) {
+                    continue;
+                }
+
+                String indexName = index.getName();
+                IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+                String backingIndexDownsamplingSource = backingIndex.getSettings().get(IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME_KEY);
+
+                // if the current index is not a downsample we want to mark the index as read-only before proceeding with downsampling
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource) == false
+                    && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
+                    // time to downsample, but first we have to mark the index as read-only
+                    affectedIndices.add(index);
+                    AddIndexBlockRequest addIndexBlockRequest = new AddIndexBlockRequest(WRITE, indexName).masterNodeTimeout(
+                        TimeValue.MAX_VALUE
+                    );
+                    transportActionsDeduplicator.executeOnce(
+                        addIndexBlockRequest,
+                        new ErrorRecordingActionListener(indexName, errorStore),
+                        (req, reqListener) -> addIndexBlock(addIndexBlockRequest, reqListener)
+                    );
+                    continue;
+                }
+
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource)
+                    && backingIndexDownsamplingStatus.equals(SUCCESS)) {
+                    // if the backing index is a downsample index itself, let's check if its source index still exists as we must delete it
+                    Map<String, String> lifecycleMetadata = backingIndex.getCustomData(LIFECYCLE_CUSTOM_INDEX_METADATA_KEY);
+                    if (lifecycleMetadata == null || lifecycleMetadata.containsKey(REPLACEMENT_SOURCE_INDEX) == false) {
+                        // this is a downsampling index that was not added by data stream lifecycle in this data stream
```

Comment:
Should we document that these kind of indices manually added to a data stream will be treated differently? (In the next PR)

Reply:
Yes ++